### PR TITLE
[TeamsDevices] [nGMS] Add DeviceCodeFlow (DCF) to Broker

### PR DESCRIPTION
**Pull request by Pallav2309**
### **What**
This PR includes adding DCF to Broker for Teams Devices' nGMS changes. Feature is controlled by `FlightNames.ENABLE_NGMS_FLOW` and is currently set to false. 

### **Why** and **How**
nGMS Teams Devices Design PR: https://identitydivision.visualstudio.com/DevEx/_git/AuthLibrariesApiReview/pullrequest/6564?_a=files

// TODO:
1) Add Telemetry once Msal to Broker telemetry is setup. Currently, I've added OTel spans for broker side 
2) Add UTs

### **Testing and Validation:**
Tested DCF` authResult ` flow and able to get the User code and other DCF info successfully
MSAL_DCF_FETCH_AUTH_RESULT

Tested DCF `acquireToken` by signing into `/devicelogin` and able to get token successfully
MSAL_DCF_ACQUIRE_TOKEN

Validated that an exception shall be thrown if nGMS Feature flag (FlightNames.ENABLE_NGMS_FLOW)
 is disabled 
MSAL_DCF_NOT_SUPPORTED_BEFORE_NGMS

### CODE REVIEW

File: msal/src/main/java/com/microsoft/identity/client/PublicClientApplication.java [hunk 1]
```diff
@@ -1834,6 +1835,45 @@ public void onError(MsalException exception) {
...
+        try {
+            final DeviceCodeFlowCommand deviceCodeFlowCommand = new DeviceCodeFlowCommand(
+                    commandParameters,
+                    MSALControllerFactory.getDefaultController(
+                            mPublicClientConfiguration.getAppContext(),
+                            commandParameters.getAuthority(),
+                            mPublicClientConfiguration
+                    ),
+                    deviceCodeFlowCommandCallback,
+                    PublicApiId.DEVICE_CODE_FLOW_WITH_CALLBACK
```
Comment (mohitc1) [inline msal/src/main/java/com/microsoft/identity/client/PublicClientApplication.java]: >DEVICE_CODE_FLOW_WITH_CALLBACK [](http://example.com/codeflow?start=32&length=30)

new API id here?
(Unless we are deprecating the below one)

Comment (Pallav2309) [reply msal/src/main/java/com/microsoft/identity/client/PublicClientApplication.java]: done, thanks!

File: msal/src/main/java/com/microsoft/identity/client/IPublicClientApplication.java [hunk 2]
```diff
@@ -89,6 +92,20 @@ void acquireToken(@NonNull final Activity activity,
...
+     *
+     * @param scopes   the desired access scopes
+     * @param callback callback object used to communicate with the API throughout the protocol
+     * @param claims claims Authentication Request parameter requests that specific Claims be returned from the UserInfo Endpoint and/or in the ID Token.
+     * @param correlationId  correlation id of this request
+     *
+     * Important: Use of this API requires setting the minimum_required_broker_protocol_version to
+     * "11.0" or higher.
+     */
+    void acquireTokenWithDeviceCode(@NonNull List<String> scopes, @NonNull final DeviceCodeFlowCallback callback, @Nullable ClaimsRequest claims, @Nullable UUID correlationId);
```
Comment (mohitc1) [inline msal/src/main/java/com/microsoft/identity/client/IPublicClientApplication.java]: >@NonNull List<String> scopes, @NonNull final DeviceCodeFlowCallback callback, @Nullable ClaimsRequest claims, @Nullable UUID correlationId [](http://example.com/codeflow?start=36&length=138)

nit: final for parameters where applicable

File: msal/src/main/java/com/microsoft/identity/client/internal/CommandParametersAdapter.java [hunk 3]
```diff
@@ -203,6 +205,39 @@ public static SilentTokenCommandParameters createSilentTokenCommandParameters(
         return commandParameters;
     }
 
+    public static DeviceCodeFlowCommandParameters createDeviceCodeFlowWithClaimsCommandParameters(
+            @NonNull final PublicClientApplicationConfiguration configuration,
+            @NonNull final OAuth2TokenCache tokenCache,
+            @NonNull final DeviceCodeFlowParameters parameters) {
```
Comment (mohitc1) [inline msal/src/main/java/com/microsoft/identity/client/internal/CommandParametersAdapter.java:217]: doc

File: msal/src/main/java/com/microsoft/identity/client/DeviceCodeFlowParameters.java [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+        public DeviceCodeFlowParameters.Builder self() {
+            return this;
+        }
+
+        public DeviceCodeFlowParameters build() {
+            return new DeviceCodeFlowParameters(this);
+        }
+    }
+
+}
```
Comment (mohitc1) [inline msal/src/main/java/com/microsoft/identity/client/DeviceCodeFlowParameters.java:43]: Documentation here. 
Also consider adding null/nonnull annotations as appropriate.

**Review by mohitc1 — APPROVED** [review]

**Review by somalaya — APPROVED** [review]

File: msal/src/main/java/com/microsoft/identity/client/DeviceCodeFlowParameters.java (hunk 4 above)
Comment (Amritkrsaini1) [reply msal/src/main/java/com/microsoft/identity/client/DeviceCodeFlowParameters.java:43]: null/non null is not applicable to constructors, same patterns are followed for other classes extending TokenParameters.

File: msal/src/main/java/com/microsoft/identity/client/internal/CommandParametersAdapter.java (hunk 3 above)
Comment (Amritkrsaini1) [reply msal/src/main/java/com/microsoft/identity/client/internal/CommandParametersAdapter.java:217]: done

File: msal/src/main/java/com/microsoft/identity/client/IPublicClientApplication.java [hunk 5]
```diff
@@ -89,6 +92,21 @@ void acquireToken(@NonNull final Activity activity,
...
+//     * Perform the Device Code Flow (DCF) protocol to allow a device without input capability to authenticate and get a new access token.
+//     * This flow is now supported in Broker as well. It also supports requesting Claims using the "claims" Request. Parameter.
+//     *
+//     * @param scopes   the desired access scopes
+//     * @param callback callback object used to communicate with the API throughout the protocol
+//     * @param claims claims Authentication Request parameter requests that specific Claims be returned from the UserInfo Endpoint and/or in the ID Token.
+//     * @param correlationId  correlation id of this request
+//     *
+//     * Important: Use of this API requires setting the minimum_required_broker_protocol_version to
+//     * "13.0" or higher.
```
Comment (somalaya) [inline msal/src/main/java/com/microsoft/identity/client/IPublicClientApplication.java:105]: we need to ensure that Teams devices sends this min required version. Did you create a task for sending this?

File: msal/src/main/java/com/microsoft/identity/client/internal/CommandParametersAdapter.java [hunk 6]
```diff
@@ -203,6 +205,46 @@ public static SilentTokenCommandParameters createSilentTokenCommandParameters(
...
+
+        final DeviceCodeFlowCommandParameters commandParameters = DeviceCodeFlowCommandParameters.builder()
+                .platformComponents(AndroidPlatformComponentsFactory.createFromContext(configuration.getAppContext()))
+                .applicationName(configuration.getAppContext().getPackageName())
+                .applicationVersion(getPackageVersion(configuration.getAppContext()))
+                .clientId(configuration.getClientId())
+                .isSharedDevice(configuration.getIsSharedDevice())
+                .redirectUri(configuration.getRedirectUri())
+                .oAuth2TokenCache(tokenCache)
+                .requiredBrokerProtocolVersion(MSAL_TO_BROKER_PROTOCOL_BROKER_DCF_WITH_CLAIMS_SUPPORT_MINIMUM_VERSION)
```
Comment (somalaya) [inline msal/src/main/java/com/microsoft/identity/client/internal/CommandParametersAdapter.java]: this should come from the PCA configuration. It should not be hardcoded.

**Review by somalaya — CHANGES_REQUESTED** [review]
Left some commnents related to min required broker protocol version.

File: msal/src/main/java/com/microsoft/identity/client/IPublicClientApplication.java (hunk 5 above)
Comment (Amritkrsaini1) [reply msal/src/main/java/com/microsoft/identity/client/IPublicClientApplication.java:105]: https://domoreexp.visualstudio.com/MSTeams/_workitems/edit/3055116

File: msal/src/main/java/com/microsoft/identity/client/internal/CommandParametersAdapter.java (hunk 6 above)
Comment (Amritkrsaini1) [reply msal/src/main/java/com/microsoft/identity/client/internal/CommandParametersAdapter.java]: removed hardcoding

**Review by somalaya — APPROVED** [review]

**Pull request merged**